From 4773889cba9c7b6c21f833d3661175f9d72fcf1e Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 5 Sep 2014 16:36:15 -0400
Subject: [infiniband] iser: Add missing target_put_sess_cmd for ImmedateData failure

Message-id: <1409934989-29941-18-git-send-email-agrover@redhat.com>
Patchwork-id: 91473
O-Subject: [RHEL7.1 PATCH e-stor 17/31] iser-target: Add missing target_put_sess_cmd for ImmedateData failure
Bugzilla: 1129387
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

From: Nicholas Bellinger <nab@linux-iscsi.org>

commit 6cc44a6fb46e1ecc1c28125aa8fa34d317aa9ea7 upstream.

This patch addresses a bug where an early exception for SCSI WRITE
with ImmediateData=Yes was missing the target_put_sess_cmd() call
to drop the extra se_cmd->cmd_kref reference obtained during the
normal iscsit_setup_scsi_cmd() codepath execution.

This bug was manifesting itself during session shutdown within
isert_cq_rx_comp_err() where target_wait_for_sess_cmds() would
end up waiting indefinately for the last se_cmd->cmd_kref put to
occur for the failed SCSI WRITE + ImmediateData descriptors.

This fix follows what traditional iscsi-target code already does
for the same failure case within iscsit_get_immediate_data().

Reported-by: Sagi Grimberg <sagig@dev.mellanox.co.il>
Cc: Sagi Grimberg <sagig@dev.mellanox.co.il>
Cc: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/isert/ib_isert.c b/drivers/infiniband/ulp/isert/ib_isert.c
index d493ad6..bcd9a1d 100644
--- a/drivers/infiniband/ulp/isert/ib_isert.c
+++ b/drivers/infiniband/ulp/isert/ib_isert.c
@@ -977,6 +977,8 @@ sequence_cmd:
 
  if (!rc && dump_payload == false && unsol_data)
   iscsit_set_unsoliticed_dataout(cmd);
+ else if (dump_payload && imm_data)
+  target_put_sess_cmd(conn->sess->se_sess, &cmd->se_cmd);
 
  return 0;
 }
-- 
1.7.1